man/man8/cpuload.8
.TH cpuload 8  "2021-01-07" "USER COMMANDS"
.SH NAME
cpuload \- Trace cpu load. Uses Linux eBPF/bcc.
.SH SYNOPSIS
.B cpuload.py [-h] [-t TIME] [-n NUMBER] [-u USAGE] [-c CPUMASK]
.SH DESCRIPTION
cpuload traces the cpu load, showing which tasks run out of cpu resource.

It display top N tasks use more than U percent cpu resource when the cpu
doesn't enter idle state for more than T ns.

This works by tracing the sched switch events using tracepoints.

Since this uses BPF, only the root user can use this tool.
.SH REQUIREMENTS
CONFIG_BPF and bcc.
.SH OPTIONS
.TP
\-h
Print usage message.
.TP
\-t TIME
Minimum nsec to print
.TP
\-n NUMBER
Maximum tasks to print
.TP
\-u USAGE
Minimum usage to print
.TP
\-c CPUMASK
Target cpus' mask to print
.SH EXAMPLES
.TP
Print top 3 tasks which use more than 30 percent of CPU in 100 ms:
#
.B cpuload -t 100000000 -n 3 -u 30
.TP
Only print tasks on cpu0 and cpu1:
#
.B cpuload -c 3
.SH FIELDS
.TP
COMM
Process name
.TP
PID
Process ID
.TP
CPU
CPU ID
.TP
TIME(ms)
Time of the task run, in millisecond
.TP
USAGE
CPU usage of the task
.SH OVERHEAD
This traces the kernel sched_switch tracepoints and prints output for each
event. As the rate of this is generally expected to be low (< 1000/s), the
overhead is also expected to be negligible.
.SH SOURCE
This is from bcc.
.IP
https://github.com/iovisor/bcc
.PP
Also look in the bcc distribution for a companion _examples.txt file containing
example usage, output, and commentary for this tool.
.SH OS
Linux
.SH STABILITY
Unstable - in development.
.SH AUTHOR
Ethercflow 


debian/changelog
bcc (0.18.0-1) unstable; urgency=low

  * Support for kernel up to 5.10
  * add bpf kfunc/kretfunc C++ example
  * add PT_REGS_PARMx_SYSCALL helper macro
  * biolatency: allow json output
  * biolatpcts: support measuring overall latencies between two events
  * fix build when ENABLE_CLANG_JIT is disabled
  * doc update and bug fixes

 -- Yonghong Song <ys114321@gmail.com>  Mon, 4 Jan 2021 17:00:00 +0000

bcc (0.17.0-1) unstable; urgency=low

  * Support for kernel up to 5.9
  * usdt: add uprobe refcnt support
  * use newer llvm/clang versions in debian packaging if possible
  * add bpf iterator C++ support
  * new bcc tools: tcprtt, netqtop, swapin, tcpsynbl, threadsnoop
  * tcpconnect: add DNS correlation to connect tracking
  * new libbpf-tools: llcstat, numamove, runqlen, runqlat, softirgs, hardirqs
  * doc update, bug fixes and some additional arguments for tools

 -- Yonghong Song <ys114321@gmail.com>  Thu, 29 Oct 2020 17:00:00 +0000

bcc (0.16.0-1) unstable; urgency=low

  * Support for kernel up to 5.8
  * trace.py: support kprobe/uprobe func offset
  * support raw perf config for perf_event_open in python
  * add BPFQueueStackTable support
  * added Ringbuf support support
  * libbpf-tools: readahead, biosnoop, bitesize, tcpconnlat, biopattern, biostacks
  * bug fixes and some additional arguments for tools

 -- Yonghong Song <ys114321@gmail.com>  Sat, 22 Aug 2020 17:00:00 +0000

bcc (0.15.0-1) unstable; urgency=low

  * Support for kernel up to 5.7
  * new tools: funcinterval.py, dirtop.py
  * support lsm bpf programs
  * support multiple pid/tids for offwaketime
  * usdt: add helpers to set semaphore values
  * turn off x86 jump table optimization during jit compilation
  * add support to use bpf_probe_read[_str_}{_user,kernel} in all bpf
  *    programs, fail back to old bpf_probe_read[_str] for old kernels
  * tools: add filtering by mount namespace
  * libbpf-tools: cpudist, syscount, execsnoop, vfsstat
  * lots of bug fixes and a few additional arguments for tools

 -- Yonghong Song <ys114321@gmail.com>  Mon, 19 Jun 2020 17:00:00 +0000

bcc (0.14.0-1) unstable; urgency=low

  * Support for kernel up to 5.6
  * new tools: biolatpcts.py
  * libbpf-tools: tools based on CORE and libbpf library directly
  * add --cgroupmap to various tools, filtering based cgroup
  * support kfunc (faster kprobe) for vfsstat, klockstat and opensnoop
  * lots of bug fixes and a few additional arguments for tools

 -- Yonghong Song <ys114321@gmail.com>  Mon, 20 Apr 2020 17:00:00 +0000

bcc (0.13.0-1) unstable; urgency=low

  * Support for kernel up to 5.5
  * bindsnoop tool to track tcp/udp bind information
  * added compile-once run-everywhere based libbpf-tools, currently
    only runqslower is implemented.
  * new map support: sockhash, sockmap, sk_storage, cgroup_storage
  * enable to run github actions on the diff
  * cgroupmap based cgroup filtering for opensnoop, execsnoop and bindsnoop.
  * lots of bug fixes.

 -- Yonghong Song <ys114321@gmail.com>  Wed, 19 Feb 2020 17:00:00 +0000

bcc (0.12.0-1) unstable; urgency=low

  * Support for kernel up to 5.4
  * klockstat tool to track kernel mutex lock statistics
  * cmake option CMAKE_USE_LIBBPF_PACKAGE to build a bcc shared library
    linking with distro libbpf_static.a
  * new map.lookup_or_try_init() API to remove hidden return in
    map.lookup_or_init()
  * BPF_ARRAY_OF_MAPS and BPF_HASH_OF_MAPS support
  * support symbol offset for uprobe in both C++ and python API,
    kprobe already has the support
  * bug fixes for trace.py, tcpretrans.py, runqslower.py, etc.

 -- Yonghong Song <ys114321@gmail.com>  Tue, 10 Dec 2019 17:00:00 +0000

bcc (0.11.0-1) unstable; urgency=low

  * Support for kernel up to 5.3
  * Corresponding libbpf submodule release is v0.0.5
  * Fix USDT issue with multi-threaded applications
  * Fixed the early return behavior of lookup_or_init
  * Support for nic hardware offload
  * Fixed and Enabled Travis CI
  * A lot of tools change with added new options, etc.

 -- Yonghong Song <ys114321@gmail.com>  Tue, 03 Oct 2019 17:00:00 +0000

bcc (0.10.0-1) unstable; urgency=low

  * Support for kernel up to 5.1
  * corresponding libbpf submodule release is v0.0.3
  * support for reading kernel headers from /proc
  * libbpf.{a,so} renamed to libcc_bpf.{a,so}
  * new common options for some tools
  * new tool: drsnoop
  * s390 USDT support

 -- Brenden Blanco <bblanco@gmail.com>  Tue, 28 May 2019 17:00:00 +0000

bcc (0.9.0-1) unstable; urgency=low

  * Adds support for BTF
  * Uses libbpf common library to wrap syscall API
  * Many bugfixes and new tools

 -- Brenden Blanco <bblanco@gmail.com>  Thu, 07 Mar 2019 17:00:00 +0000

bcc (0.8.0-1) unstable; urgency=low

  * Support for kernel up to 5.0

 -- Brenden Blanco <bblanco@gmail.com>  Fri, 11 Jan 2019 17:00:00 +0000

bcc (0.7.0-1) unstable; urgency=low

  * Support for kernel up to 4.18

 -- Brenden Blanco <bblanco@gmail.com>  Tue, 04 Sep 2018 17:00:00 +0000

bcc (0.6.1-1) unstable; urgency=low

  * Build support for Fedora 28 and Ubuntu 18.04
  * Add option to change license
  * Optimizations for some uses of bpf_probe_reads

 -- Brenden Blanco <bblanco@gmail.com>  Mon, 23 Jul 2018 17:00:00 +0000

bcc (0.6.0-1) unstable; urgency=low

  * Support for kernel up to 4.17
  * Many bugfixes
  * Many new tools
  * Improved python3 support

 -- Brenden Blanco <bblanco@gmail.com>  Wed, 13 Jun 2018 17:00:00 +0000

bcc (0.5.0-1) unstable; urgency=low

  * Support for USDT in ARM64
  * Bugfixes for 4.14 in some tools
  * Fixes for smoke test failures
  * Runtime memory usage reductions

 -- Brenden Blanco <bblanco@gmail.com>  Wed, 29 Nov 2017 17:00:00 +0000

bcc (0.4.0-1) unstable; urgency=low

  * Bugfixes
  * Support for kernel up to 4.14

 -- Brenden Blanco <bblanco@gmail.com>  Fri, 20 Oct 2017 17:00:00 +0000

bcc (0.3.0-1) unstable; urgency=low

  * Many bugfixes
  * Many tools converted to perf ring buffer
  * New utilities in tools/

tools/cpuload.py
#!/usr/bin/python
# @lint-avoid-python-3-compatibility-imports
#
# cpuload   Display top N tasks use more than U percent cpu resource when
#           the cpu doesn't enter idle state for more than T ns.
#
# USAGE: cpuload [-h] [-t time] [-n number] [-u usage] [-c cpumask]
#
# This uses in-kernel eBPF maps to cache task details (PID and comm) by
# sched_switch, as well as a starting timestamp for calculating cpu usage.
#
# Licensed under the Apache License, Version 2.0 (the "License")
# Copyright (C) 2021 Liu Chao.
#
# 07-Jan-2021    Liu Chao       Created this.

from __future__ import print_function
from bcc import BPF
from bcc.utils import printb
import argparse
from datetime import datetime

# arguments
examples = """examples:
    ./cpuload                # display tasks when cpu overload
    ./cpuload -t 100000000   # display when non-idle for 100000000 ns
    ./cpuload -n 5           # display top 5 tasks details
    ./cpuload -u 30          # display tasks use more than 30 percent cpu
    ./cpuload -c 3           # only display cpu0 and cpu1
"""
parser = argparse.ArgumentParser(
    description="display tasks when cpu overload",
    formatter_class=argparse.RawDescriptionHelpFormatter,
    epilog=examples)
parser.add_argument("-t", "--time", default=1000000000,
    help="minimum nsec to print, default 1000000000")
parser.add_argument("-n", "--number", default=3,
    help="maximum tasks to print, default 3")
parser.add_argument("-u", "--usage", default=30,
    help="minimum usage to print, default 30")
parser.add_argument("-c", "--cpumask", default=-1,
    help="target cpus' mask to print, default all cpus")
parser.add_argument("--ebpf", action="store_true",
    help=argparse.SUPPRESS)
args = parser.parse_args()
time = int(args.time)
number = int(args.number)
usage = int(args.usage)
target = int(args.cpumask)
debug = 0

# define BPF program
bpf_text = """
#include <linux/sched.h>

#define WARN_NSEC """ + str(time) + """
#define MAX_ENTRY 1024

enum state {
    IDLE,
    OVERLOAD,
    OVERFLOW
};

struct cpu_data_t {
    u64 begin_time;
    u64 prev_time;
    int number;
    int state;
};

struct task_data_t {
    u32 pid;
    char comm[TASK_COMM_LEN];
    u64 delta;
};

struct data_t {
    u64 total_time;
    int number;
};

BPF_PERCPU_ARRAY(cpu_data, struct cpu_data_t, 1);

BPF_PERCPU_ARRAY(task_data, struct task_data_t, MAX_ENTRY);

BPF_PERF_OUTPUT(events);
TRACEPOINT_PROBE(sched, sched_switch) {
    u32 zero = 0;
    u64 now = bpf_ktime_get_ns();
    struct data_t data = {};
    struct cpu_data_t *cpu = cpu_data.lookup(&zero);
    struct task_data_t *task;

    if (cpu == 0)
        return 0;

    if (args->prev_pid == 0 || cpu->begin_time == 0) {
        cpu->begin_time = now;
        cpu->prev_time = now;
        cpu->number = 0;
        cpu->state = IDLE;
        return 0;
    }

    if (cpu->state == OVERLOAD || cpu->state == OVERFLOW)
        return 0;

    if (cpu->number >= MAX_ENTRY) {
        data.number = MAX_ENTRY;
        data.total_time = cpu->prev_time - cpu->begin_time;
        cpu->state = OVERFLOW;
        events.perf_submit(args, &data, sizeof(data));
        return 0;
    }

    task = task_data.lookup(&cpu->number);
    if (task == 0)
        return 0;

    task->pid = args->prev_pid;
    __builtin_memcpy(&task->comm, &args->prev_comm, sizeof(task->comm));
    task->delta = now - cpu->prev_time;
    cpu->prev_time = now;
    cpu->number++;

    data.total_time = now - cpu->begin_time;
    if (data.total_time > WARN_NSEC) {
        data.number = cpu->number;
        cpu->state = OVERLOAD;
        events.perf_submit(args, &data, sizeof(data));
    }
    return 0;
}
"""

if debug or args.ebpf:
    print(bpf_text)
    if args.ebpf:
        exit()

# initialize BPF
b = BPF(text=bpf_text)

print("Tracing task switch. Output when cpu is overload. Ctrl-C to end.")

print("%-19s %-14s %-7s %-4s %-8s %-5s" %
        ("DATE", "COMM", "PID", "CPU", "TIME(ms)", "USAGE"))

# process event
def print_event(cpu, data, size):
    if target & 1 << cpu == 0:
        return 0
    date = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
    data = b["events"].event(data)
    dic = {}
    for i in range(0, data.number):
        task = b["task_data"][i][cpu]
        entry = dic.get(task.pid)
        if entry is not None:
            entry.delta += task.delta
        else:
            dic[task.pid] = task

    count = 0
    for item in sorted(dic.items(), key=lambda x: x[1].delta, reverse=True):
        if count >= number:
            break
        task = item[1]
        u = task.delta * 100 / data.total_time
        if u < usage:
            break
        print("%s %-14.14s %-7s %-4s %-8.3f %05.2f%%" % (
            date,
            task.comm.decode("utf-8", "replace"),
            task.pid,
            cpu,
            float(task.delta) / 1000000,
            u))
        count += 1
    dic.clear()

# loop with callback to print_event
b["events"].open_perf_buffer(print_event)
while 1:
    try:
        b.perf_buffer_poll()
    except KeyboardInterrupt:
        exit()
